docs/blockentities/container.md

```diff
@@ -0,0 +1,250 @@
+# Containers
+
+A popular use case of [block entities][blockentity] is to store items of some kind. Some of the most essential [blocks][block] in Minecraft, such as the furnace or the chest, use block entities for this purpose. To store items on something, Minecraft uses `Container`s.
+
+The `Container` interface defines methods such as `#getItem`, `#setItem` and `#removeItem` that can be used to query and update the container. Since it is an interface, it does not actually contain a backing list or other data structure, that is up to the implementing system.
+
+Due to this, `Container`s can not only be implemented on block entities, but any other class as well. Notable examples include entity inventories, as well as common modded [items][item] such as backpacks.
+
+## Basic Container Implementation
+
+Containers can be implemented in any way you like, so long as you satisfy the dictated methods (as with any other interface in Java). However, it is common to use a `NonNullList<ItemStack>` with a fixed length as a backing structure. Single-slot containers may also simply use an `ItemStack` field instead.
+
+For example, a basic implementation of `Container` with a size of 27 slots (one chest) could look like this:
+
+```java
+public class MyContainer implements Container {
+    private final NonNullList<ItemStack> items = NonNullList.withSize(
+            // The size of the list, i.e. the amount of slots in our container.
+            27,
+            // The default value to be used in place of where you'd use null in normal lists.
+            ItemStack.EMPTY
+    );
+
+    // The amount of slots in our container.
+    @Override
+    public int getContainerSize() {
+        return 27;
+    }
+
+    // Whether the container is considered empty.
+    @Override
+    public boolean isEmpty() {
+        return items.stream().allMatch(ItemStack::isEmpty);
+    }
+
+    // Return the item stack in the specified slot.
+    @Override
+    public ItemStack getItem(int slot) {
+        return items.get(slot);
+    }
+
+    // Remove the specified amount of items from the given slot, returning the stack that was just removed.
+    // We defer to ContainerHelper here, which does this as expected for us.
+    @Override
+    public ItemStack removeItem(int slot, int amount) {
+        return ContainerHelper.removeItem(items, slot, amount);
+    }
+
+    // Remove all items from the specified slot, returning the stack that was just removed.
+    // We again defer to ContainerHelper here.
+    @Override
+    public ItemStack removeItemNoUpdate(int slot) {
+        return ContainerHelper.takeItem(items, slot);
+    }
+
+    // Set the given item stack in the given slot.
+    @Override
+    public void setItem(int slot, ItemStack stack) {
+        items.set(slot, stack);
```

Comment:
Should respect the max stack size that can be accepted by the container. `stack.limitSize(this.getMaxStackSize(stack));`